### PR TITLE
Feature/add headers to commands requests

### PR DESCRIPTION
**Pull request by diogozanattotake**
Create a method to add Bot Id and State Id to Header requests when URI is "http.msging.net" or another internal URI

### CODE REVIEW

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 1]
```diff
@@ -50,8 +53,16 @@ public override async Task ExecuteAsync(IContext context, ProcessHttpSettings se
                             contentType ?? "application/json");
                     }
 
-                    AddUserToHeaders(httpRequestMessage, context);
-                    AddBotIdentityToHeaders(httpRequestMessage, context);
+                    if (!_configuration.Internal_Uris.Contains(settings.Uri.AbsoluteUri))
```
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:56]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 2]
```diff
@@ -121,14 +132,23 @@ private void AddUserToHeaders(HttpRequestMessage httpRequestMessage, IContext co
         /// </summary>
         /// <param name="httpRequestMessage"></param>
         /// <param name="context"></param>
-        private void AddBotIdentityToHeaders(HttpRequestMessage httpRequestMessage, IContext context)
+        private void AddBotIdentityToHeaders(HttpRequestMessage httpRequestMessage, IContext context, string uri)
```
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:135]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 3]
```diff
@@ -121,14 +132,23 @@ private void AddUserToHeaders(HttpRequestMessage httpRequestMessage, IContext co
...
         {
             if (context.Flow.ConfigurationFlagIsEnabled(ADD_BOT_KEY))
             {
                 httpRequestMessage.Headers.Add(Constants.BLIP_BOT_HEADER, context.OwnerIdentity);
             }
         }
 
+        private void AddHeadersToCommandRequest(HttpRequestMessage httpRequestMessage, string currentStateId, string ownerIdentity, string uri, IContext context)
```
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:143]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 4]
```diff
@@ -121,14 +132,23 @@ private void AddUserToHeaders(HttpRequestMessage httpRequestMessage, IContext co
...
+        {
+            if (uri.Contains(_configuration.Internal_Uris))
```
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:145]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:56]: ok

File: src/Take.Blip.Builder/FlowManager.cs [hunk 5]
```diff
@@ -545,11 +546,15 @@ private async Task ProcessActionsAsync(LazyInput lazyInput, IContext context, Ac
                     {
                         JObject jObjectSettings = null;
                         var stringifySetting = stateAction.Settings?.ToString(Formatting.None);
-                        
+
                         if (stringifySetting != null)
                         {
                             stringifySetting = await _variableReplacer.ReplaceAsync(stringifySetting, context, cancellationToken);
                             jObjectSettings = JObject.Parse(stringifySetting);
+                            if (action.Type == ACTION_PROCESS_HTTP)
```
[the remark this answers is not in the document]
Comment (diogozanattotake) [reply src/Take.Blip.Builder/FlowManager.cs:554]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 6]
```diff
@@ -50,8 +53,16 @@ public override async Task ExecuteAsync(IContext context, ProcessHttpSettings se
                             contentType ?? "application/json");
                     }
 
+                    if (settings.Uri.AbsoluteUri.Contains(_configuration.InternalUris))
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:56]: move to a new method and add the rule of ";"

File: src/Take.Blip.Builder/FlowManager.cs [hunk 7]
```diff
@@ -710,6 +712,18 @@ private async Task<bool> ValidateInputAsync(Message message, State state, LazyIn
             return true;
         }
 
+        private JObject AddStateIdToSettings(string actionType, JObject jObjectSettings, string stateId)
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder/FlowManager.cs:715]: this method doesn't need to return a value...

Comment (diogozanattotake) [reply src/Take.Blip.Builder/FlowManager.cs]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs (hunk 6 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 8]
```diff
@@ -140,10 +133,17 @@ private void AddBotIdentityToHeaders(HttpRequestMessage httpRequestMessage, ICon
...
-             httpRequestMessage.Headers.Add(Constants.BLIP_BOT_HEADER, ownerIdentity);
-             httpRequestMessage.Headers.Add(Constants.BLIP_STATEID_HEADER, currentStateId);
+        private void AddHeadersToCommandRequest(HttpRequestMessage httpRequestMessage, string currentStateId, string ownerIdentity, string absoluteUri)
+        {
+            var uriList = _configuration.InternalUris.Split(";");
+            foreach (var uri in uriList)
+            {
+                if (absoluteUri.Contains(uri))
+                {
+                    httpRequestMessage.Headers.Add(Constants.BLIP_BOT_HEADER, ownerIdentity);
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:143]: the headers will be duplicated if two uris were contained in absoluteUris

Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs]: ok

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs [hunk 9]
```diff
@@ -129,6 +134,24 @@ private void AddBotIdentityToHeaders(HttpRequestMessage httpRequestMessage, ICon
             }
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:134]: if "CheckInternalUris(absoluteUri)" is true, the BLIP_BOT_HEADER will be duplicated

Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:134]: is good if you add a unit test for it...

File: src/Take.Blip.Builder/FlowManager.cs [hunk 10]
```diff
@@ -710,6 +712,14 @@ private async Task<bool> ValidateInputAsync(Message message, State state, LazyIn
             return true;
         }
 
+        private void AddStateIdToSettings(string actionType, JObject jObjectSettings, string stateId)
+        {
+            if (actionType == ACTION_PROCESS_HTTP)
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder/FlowManager.cs:717]: replace with
if (actionType != ACTION_PROCESS_HTTP) {
  return;
}
...

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 11]
```diff
@@ -309,5 +312,101 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+            HttpRequestMessage requestMessage = null;
+            HttpClient
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            // Act
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            // Assert
+            requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeTrue();
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:352]: why do header contain X-Blip-Bot e X-StateId if URI doesen't contain "msging.net"?

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 12]
```diff
@@ -309,5 +312,101 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+            };
+
+            HttpRequestMessage requestMessage = null;
+            HttpClient
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            if(!settings.Uri.AbsolutePath.Contains("https://blip.ai") && Context.Flow.ConfigurationFlagIsEnabled(botIdentifierConfigVariableName))
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:397]: why this condition if "Uri" and  "Context.Flow.ConfigurationFlagIsEnabled" is a constant value?

File: src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs (hunk 9 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder/Actions/ProcessHttp/ProcessHttpAction.cs:134]: Done

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 11 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:352]: Done, change the URI from "https://blip.ai" to "http.msging.net"

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 12 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs]: removed this test, at this moment this is an unnecessary

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 13]
```diff
@@ -309,5 +313,49 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+            HttpClient
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            // Act
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            // Assert
+            requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeTrue();
+            requestMessage.Headers.Contains("X-Blip-StateId").ShouldBeTrue();
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:354]: this souldn't be true when URI not matches with internal URI...
is this test checked with success?

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 14]
```diff
@@ -177,7 +180,7 @@ public async Task ProcessActionWithBotIdentifierHeaderShouldSucceed()
 
             var settings = new ProcessHttpSettings
             {
-                Uri = new Uri("https://blip.ai"),
+                Uri = new Uri("http.msging.net"),
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:183]: this is no valid URI. Replace with "https://test.msging.net" for example.

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 15]
```diff
@@ -309,5 +313,49 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
                     h => h.RequestUri.Equals(settings.Uri)), Arg.Any<CancellationToken>());
         }
+
+        [Fact]
+        public async Task ProcessAction_AddBotIdentifierAndStateId()
+        {
+
+            var settings = new ProcessHttpSettings
+            {
+                Uri = new Uri("https://blip.ai"),
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:323]: should'd this host be "https://test.msging.net"?

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 14 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:183]: DOne

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 15 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:323]: Done

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 13 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs]: Done

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 16]
```diff
@@ -309,5 +313,57 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+            if (settings.Uri.AbsoluteUri.Contains("msging.net"))
+            {
+                requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeTrue();
+                requestMessage.Headers.Contains("X-Blip-StateId").ShouldBeTrue();
+            }
+            else
+            {
+                requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeFalse();
+                requestMessage.Headers.Contains("X-Blip-StateId").ShouldBeFalse();
+            }
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:362]: split this "if" into two test

Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:404]: Done

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 17]
```diff
@@ -309,5 +313,99 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+            HttpRequestMessage requestMessage = null;
+            HttpClient
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            // Act
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            // Assert
+            if (settings.Uri.AbsoluteUri.Contains("msging.net"))
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:353]: unnecessary "if".
settings.Uri.AbsoluteUri.Contains will always be true...

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 18]
```diff
@@ -309,5 +313,99 @@ public async Task ProcessAction_CheckConfigurationVariableValues(string botIdent
...
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            // Act
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            // Assert
+            if (!settings.Uri.AbsoluteUri.Contains("msging.net"))
+            {
+                requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeFalse();
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:402]: this block code is never called...

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 17 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:353]: Done

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs (hunk 18 above)
Comment (diogozanattotake) [reply src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs]: DOne

File: src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs [hunk 19]
```diff
@@ -397,15 +441,106 @@ public async Task ProcessAction_CheckIfInternalUriNotMatch()
...
+            HttpRequestMessage requestMessage = null;
+            HttpClient
+                .SendAsync(Arg.Do<HttpRequestMessage>(m => requestMessage = m), Arg.Any<CancellationToken>())
+                .ReturnsForAnyArgs(httpResponseMessage);
+
+            // Act
+            await target.ExecuteAsync(Context, JObject.FromObject(settings), CancellationToken);
+
+            // Assert
+            requestMessage.Headers.Contains("X-Blip-Bot").ShouldBeTrue();
```
Comment (rafael-santiago-take) [inline src/Take.Blip.Builder.UnitTests/Actions/ProcessHttpActionTests.cs:492]: it doesen't verify if exists a duplicate header...

**Review by rafael-santiago-take — APPROVED** [review]

**Review by DanielFonsecaTake — APPROVED** [review]

**Review by fadoaglauss — APPROVED** [review]

**Review by Guilhermealvestake — APPROVED** [review]

**Pull request merged**